src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -6963,16 +6964,21 @@ private bool IsPossibleNewExpression()
             return null;
         }
 
+        private BlockSyntax ParsePossiblyAttributedBlock(bool isMethodBody = false, bool isAccessorBody = false)
+            => ParseBlock(this.ParseAttributeDeclarations(), isMethodBody, isAccessorBody);
+
         // If "isMethodBody" is true, then this is the immediate body of a method/accessor.
         // In this case, we create a many-child list if the body is not a small single statement.
         // This then allows a "with many weak children" red node when the red node is created.
         // If "isAccessorBody" is true, then we produce a special diagnostic if the open brace is
         // missing.  Also, "isMethodBody" must be true.
-        private BlockSyntax ParseBlock(bool isMethodBody = false, bool isAccessorBody = false)
+        private BlockSyntax ParseBlock(SyntaxList<AttributeListSyntax> attributes, bool isMethodBody = false, bool isAccessorBody = false)
         {
             // Check again for incremental re-use, since ParseBlock is called from a bunch of places
             // other than ParseStatementCore()
-            if (this.IsIncrementalAndFactoryContextMatches && this.CurrentNodeKind == SyntaxKind.Block)
+            if (this.IsIncrementalAndFactoryContextMatches &&
+                this.CurrentNodeKind == SyntaxKind.Block &&
+                attributes.Count == 0)
```

Comment:
>attributes.Count == 0 [](start = 16, length = 21)

This condition is a bit subtle but I believe correct.  I think it depends on the fact that the node about to be eaten could not possibly have any attributes on it, otherwise we would have crumbled the block in order to parse those attributes.

Reply:
Agreed. Another way to think about it is this:

If we have parsed attributes, then we've already consumed a portion of the text and produced nodes that we *must* do something with.  If we then parse out a block from the previous tree (and return it), that means we would end up 'losing' those attributes leading to a final tree that didn't represent the original source.  (This is actually the error i caused in incremental parsing when making the change to add attributes/modifiers to a namespace decl).

Now, technically, we could try to somehow take the attributes we parsed out so far, along with the old block and merge them somehow.  But that just feels like a lot of complexity (and potentially subtle bugs) for an enormously unlikely (and possibly impossible) scenario.

So, generally, if we've consumed anything so far into temporary parser state, we basically cannot consume any old nodes from the tree. 

In general, we should only

Reply:
This comment ends in the middle of a sentence, but I think you have conveyed the thought you meant to convey :smile: